include/spblas/algorithms/detail/spgemm/spgemm_innerproduct.hpp

```diff
@@ -0,0 +1,142 @@
+#pragma once
+
+#include <spblas/backend/backend.hpp>
+#include <spblas/concepts.hpp>
+#include <spblas/detail/log.hpp>
+
+#include <spblas/algorithms/detail/sparse_dot_product.hpp>
+#include <spblas/algorithms/transposed.hpp>
+#include <spblas/backend/csr_builder.hpp>
+#include <spblas/backend/spa_accumulator.hpp>
+#include <spblas/detail/operation_info_t.hpp>
+
+namespace spblas {
+
+// C = AB
+// CSR * CSC -> CSR
+// SpGEMM (Inner Product)
+template <matrix A, matrix B, matrix C>
+  requires(__backend::row_iterable<A> && __backend::column_iterable<B> &&
+           __detail::is_csr_view_v<C>)
+void multiply(A&& a, B&& b, C&& c) {
+  log_trace("");
+  if (__backend::shape(a)[0] != __backend::shape(c)[0] ||
+      __backend::shape(b)[1] != __backend::shape(c)[1] ||
+      __backend::shape(a)[1] != __backend::shape(b)[0]) {
+    throw std::invalid_argument(
+        "multiply: matrix dimensions are incompatible.");
+  }
+
+  using T = tensor_scalar_t<C>;
+  using I = tensor_index_t<C>;
+
+  __backend::spa_accumulator<T, I> dot_product_acc(__backend::shape(a)[1]);
+  __backend::spa_accumulator<T, I> c_row(__backend::shape(c)[1]);
+  __backend::csr_builder c_builder(c);
+
+  for (auto&& [i, a_row] : __backend::rows(a)) {
+    c_row.clear();
+
+    if (!__ranges::empty(a_row)) {
+      for (auto&& [j, b_column] : __backend::columns(b)) {
+        if (!__ranges::empty(b_column)) {
+          auto v =
+              __detail::sparse_dot_product<T>(dot_product_acc, a_row, b_column);
+
+          if (v.has_value()) {
+            c_row[j] += v.value();
+          }
+        }
+      }
+      c_row.sort();
```

Comment:
very nice, double for loop over nonempty rows of A and B then sparse dot product of resulting row,col pair

Reply:
so you could separate this multiply_compute == sparse_dot_product of patterns to determine if they intersect, to compute the nnz pattern.  Then multiply_fill == sparse_dot_product of pattern + values  to fill the values... The second pass could be done only over sparse pattern ...  since we need user to allocate C , it can more naturally fit into the two APIs ...